MM-42489_Revamp: Login screen webapp

julmondragon opened the pull request:
#### Summary
Apply new designs for the Login screen which should consider also cases when:
- Default login
- OAuth options are available
- Custom branding is configured
- MFA token

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42489

#### Design
https://www.figma.com/file/7BM8S6BL597zBE83U6H8CK/MM-29159-Desktop-%26-Webapp-Login-Update

#### Screenshots
##### Login with username/email
<img width="1363" alt="Screen Shot 2022-05-09 at 12 41 16 PM" src="https://user-images.githubusercontent.com/79058848/167467913-84003cf4-0a3f-4f8a-a2d4-1e72f7eaf30b.png">

##### Login with username/email and external options
<img width="1363" alt="Screen Shot 2022-05-09 at 12 40 38 PM" src="https://user-images.githubusercontent.com/79058848/167468002-29c90af9-6fcf-48c9-9140-c17376e5dd40.png">

##### Login with external options
<img width="1364" alt="Screen Shot 2022-05-09 at 12 45 21 PM" src="https://user-images.githubusercontent.com/79058848/167468078-2b0f9848-19f0-4636-af7a-6c73abe27c8e.png">

##### MFA enabled
![mfa](https://user-images.githubusercontent.com/79058848/167461814-0096c4d0-9e9c-4530-8b9c-70eed6927586.gif)

##### Messages
###### Email verified
<img width="1363" alt="Screen Shot 2022-05-09 at 12 55 44 PM" src="https://user-images.githubusercontent.com/79058848/167469415-1d5c008a-8be9-4d15-9466-fa2571ea8f68.png">

###### Password updated
<img width="1365" alt="Screen Shot 2022-05-09 at 12 57 00 PM" src="https://user-images.githubusercontent.com/79058848/167469472-f1ad64a3-aa8d-4c69-ac96-4befc368c284.png">

###### Session expired
<img width="1209" alt="Screen Shot 2022-05-09 at 12 19 28 PM" src="https://user-images.githubusercontent.com/79058848/167463292-87947b9f-e38a-4e6c-a907-dd0e4627ffc4.png">

###### Failed credentials
![failure](https://user-images.githubusercontent.com/79058848/167462643-ea992bd9-f2b6-4f1a-859e-3e2515a3dd09.gif)

###### No sign-in methods
<img width="1368" alt="Screen Shot 2022-05-09 at 1 00 13 PM" src="https://user-images.githubusercontent.com/79058848/167469795-84416234-18d4-480e-b217-c43e8b076843.png">

##### Responsive
###### Default
![May-05-2022 16-08-32](https://user-images.githubusercontent.com/79058848/167214129-a3f75ad1-109c-4b5f-8efa-c721984ae859.gif)

###### Custom branding
![May-05-2022 15-52-53](https://user-images.githubusercontent.com/79058848/167214163-c53b4ca0-d17c-4a08-b613-5cd0db253cfa.gif)

#### Release Note
```release-note
Apply new designs for the Login screen:
- Default login
- OAuth options
- Custom branding
- MFA token
```

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/191574

julmondragon commented:
Currently working on update e2e.

matthewbirtch commented:
@julmondragon I will take a closer look tomorrow likely but at first glance, I noticed the background visual is pretty low-res. I'm pretty positive this background already exists in the onboarding screens and we may be able to just use that. I think it may have even been implemented as an SVG.

I'll review everything else tomorrow in more detail though

julmondragon commented:
> @julmondragon I will take a closer look tomorrow likely but at first glance, I noticed the background visual is pretty low-res. I'm pretty positive this background already exists in the onboarding screens and we may be able to just use that. I think it may have even been implemented as an SVG.
> 
> I'll review everything else tomorrow in more detail though

Login is reusing the same one onboarding process uses currently, it's in `jpg` format:(https://github.com/mattermost/mattermost-webapp/blob/master/images/admin-onboarding-background.jpg)
Just let us know if it needs to be updated. Thanks!

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/192412

pvev commented on sass/utils/_mixins.scss:
```diff
@@ -126,23 +123,71 @@
 @mixin primary-button {
```
not completely sure, but I think this mixin is used in other places in the webapp, did you verified these changes won't affect the look of the button in other places where it is used?

pvev commented on components/widgets/inputs/password_input/password_input.tsx:
```diff
@@ -0,0 +1,66 @@
+    inputSize?: SIZE;
+};
+
+const PasswordInput = React.forwardRef(({className, value, onChange, onBlur, onFocus, hasError, error, createMode, disabled, inputSize}: PasswordInputProps, ref?: React.Ref<HTMLInputElement>) => {
```
nit: place the variables in new lines so it is easier to read. Wondering why we don't have the maximum characters restriction per line in our linter.

pvev commented on components/widgets/inputs/input/input.stories.tsx:
```diff
@@ -242,6 +242,24 @@ storiesOf('Widgets/Inputs/Input', module).
+            );
+        },
+    ).
+    add(
```
nice!

pvev commented on components/widgets/icons/login_gitlab_icon.tsx:
```diff
@@ -0,0 +1,65 @@
+import {useIntl} from 'react-intl';
+
+export default function LockCircleSolidIcon(props: React.HTMLAttributes<HTMLSpanElement>) {
+    const {formatMessage} = useIntl();
```
nit: wondering if for all these icons instead of formating the aria-label message it sould make more sense to pass that value as a prop and avoid the extra import and also they would be more "customizable"

pvev commented on components/login/login.scss:
```diff
@@ -0,0 +1,384 @@
+    }
+}
+
+@media screen and (max-width: 1199px) {
```
are these breakpoints use in the media queries constant across the application? I mean, are those the ones we use and should keep using? if so, wouldn't it make sense to have it as css variables?

pvev commented on components/login/login.scss:
```diff
@@ -0,0 +1,384 @@
+            justify-content: center;
+
+            .login-body-no-login-title {
+                color: #1e325c;
```
do we have a css variable for this? if not there are several occurrences of this hex in the PR too, maybe makes sense to create one

pvev reviewed (approved):
a couple of suggestions and questions, but in general really Nice work @julmondragon :shipit:

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/192805

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/192833

JulienTant commented:
Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!

![image](https://user-images.githubusercontent.com/785518/167743389-70c8f8b7-78cd-4ae3-aba9-9252d1b6358f.png)

julmondragon commented:
> Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!
> 
Relying on this on @matthewbirtch , but the scope of this work does not include deprecation or addition of features/options currently supported.

matthewbirtch commented:
> Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!

Allowing these button colors to be customizable opens things up to a compromised design and could place emphasis on other buttons that weren't intended to have the emphasis.

Maybe we allow a custom color, but it only applies the colors to the text and button outline rather than the background?

matthewbirtch commented:
Sorry for the duplicate message, not sure how that happened

marianunez commented on components/alert_banner.tsx:
```diff
@@ -6,54 +6,53 @@ import classNames from 'classnames';
+    <div className={classNames('AlertBanner', mode, className, `AlertBanner--${variant}`)}>
+        <div className='AlertBanner__icon'>
+            <i
+                className={classNames({
+                    'icon-alert-outline': mode === 'danger' || mode === 'warning',
+                    'icon-check': mode === 'success',
+                    'icon-alert-circle-outline': mode === 'info',
+                })}
```
Nice refactor 💯

marianunez commented on components/login/login_controller/login_controller.test.jsx:
```diff
@@ -1,159 +0,0 @@
-import LocalStorageStore from 'stores/local_storage_store';
-import LoginController from 'components/login/login_controller/login_controller';
-
-describe('components/login/LoginController', () => {
```
Did we move all this test cases for the loginController somewhere ?

marianunez replied:
Perfect, I had missed it with some many files

marianunez commented on components/preparing_workspace/preparing_workspace.scss:
```diff
@@ -53,12 +53,14 @@
 
     .tertiary-button {
         @include tertiary-button;
+        @include button-medium;
```
Are these changes intended here?

julmondragon replied:
Yes, it is needed to specify styling according to size as well. As mentioned in other comments, for the next iteration of this (Signup) plan to review the option to have a button wrapper component, so that type and sizes could be set via props.

julmondragon replied:
Yes, the styles for mixin `tertiary-button` were updated, so need the size ones added here.

marianunez replied:
Gotcha, thanks for clarifying!

marianunez reviewed (commented):
Awesome work @julmondragon! That was a lot of cases to cover 🚀

julmondragon commented on components/alert_banner.scss:
```diff
@@ -1,46 +1,49 @@
     display: flex;
     overflow: hidden;
-    padding: 16px;
+    align-items: flex-start;
+    padding: 14px;
```
Yes, ensured that current use cases with different modes available continued working as expected.  Adding storybook could help.

julmondragon commented on components/login/login.tsx:
```diff
@@ -0,0 +1,806 @@
+// See LICENSE.txt for license information.
+
+/* eslint-disable max-lines */
+import React, {useState, useEffect, useRef, useCallback} from 'react';
```
Correct, basically this is the same login_controller rewritten in TS, all workflows and logic are mainly unchanged.
Improvements were contained to the render elements and more clear validations for each use case.

JulienTant commented:
> Relying on this on @matthewbirtch , but the scope of this work does not include deprecation or addition of features/options currently supported.

👍🏻 yep I was just opening the discussion!

Phrynobatrachus commented on e2e/cypress/tests/integration/system_console/site_configuration/link_customization_e20_1_spec.js:
```diff
@@ -10,6 +10,7 @@
 // Stage: @prod
 // Group: @system_console @enterprise @e20_only @not_cloud
 
+import * as TIMEOUTS from '../../fixtures/timeouts';
```
```suggestion
import * as TIMEOUTS from '../../../fixtures/timeouts';
```

julmondragon replied:
Updated!

Phrynobatrachus commented on e2e/cypress/tests/integration/system_console/site_configuration/link_customization_cloud_spec.js:
```diff
@@ -10,6 +10,7 @@
 // Stage: @prod
 // Group: @system_console @enterprise @cloud_only
 
+import * as TIMEOUTS from '../../fixtures/timeouts';
```
```suggestion
import * as TIMEOUTS from '../../../fixtures/timeouts';
```

julmondragon replied:
Updated!

julmondragon commented:
/e2e-test

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/193372

julmondragon commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/193632

julmondragon commented:
> Great work on this @julmondragon. I didn't find a whole lot. A few comments beyond the minor comments I made on code:
> ...

1. Looks like this is not possible in the cloud spinwick. We need to ensure this is tested in self managed as part of QA work. finally, there are e2e test that were updated covering this.
2. Per discussion with the team this will be worked separately. Main ticket is https://mattermost.atlassian.net/browse/MM-23853, will create specific ticket if needed.
3. Updated styles to handle larger images.
5. Updated handling of Custom Description per our review.

cc @matthewbirtch

Phrynobatrachus reviewed (approved):
https://automation-dashboard.vercel.app/cycles/68279927-588d-4e26-ab37-b4cf1d587740

Latest run seems to be solid, thanks!

marianunez reviewed (approved):
Thanks @julmondragon - great work 🚀

JulienTant reviewed (approved):
LGTM @julmondragon good job!!

stevemudie reviewed (approved):
I'll test this post merge.
